### PR TITLE
feat(storage): Add support for S3 acceleration mode

### PR DESCRIPTION
**Pull request by sdhuka**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #2190*

*Description of changes:*
* Add support for S3 accelerate endpoint
* Upgrade kotlin sdk to 0.21.1-beta
* Replace EndpointResolver with EndpointProvider as part of this [breaking change](https://github.com/awslabs/aws-sdk-kotlin/discussions/761)

*How did you test these changes?*
* Updated Unit Tests
* Added Intgeration Tests
* Manual tested custom endpoint and add logs

*Documentation update required?*
- [ ] No
- [x] [Yes](https://github.com/aws-amplify/docs/pull/5214)

*General Checklist*
- [x] Added Unit Tests
- [x] Added Integration Tests
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSEndpointProvider.kt [hunk 1]
```diff
@@ -12,16 +12,14 @@
...
 
-import aws.sdk.kotlin.runtime.endpoint.AwsEndpoint
-import aws.sdk.kotlin.runtime.endpoint.AwsEndpointResolver
-import aws.sdk.kotlin.runtime.endpoint.CredentialScope
-import aws.smithy.kotlin.runtime.http.endpoints.Endpoint
+import aws.sdk.kotlin.services.cognitoidentityprovider.endpoints.EndpointParameters
+import aws.sdk.kotlin.services.cognitoidentityprovider.endpoints.EndpointProvider
+import aws.smithy.kotlin.runtime.client.endpoints.Endpoint
 
-internal class AWSEndpointResolver(val endpoint: Endpoint) : AwsEndpointResolver {
```
Comment (sdhuka) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSEndpointProvider.kt:23]: Tested that auth api calls are hitting custom endpoint
```
System.err                          com.example.amplifyandroidpentestapp           W  [OkHttp https://my-proxy-endpoint.customerdomain.com/...] TRACE aws.smithy.kotlin.runtime.http.engine.okhttp.HttpEngineEventListener - Cognito Identity Provider-InitiateAuth-7cb66723-160d-43d7-b0bc-75080e5c2f58/Attempt-1/HTTP: call failed
System.err                          com.example.amplifyandroidpentestapp           W  java.net.ConnectException: Failed to connect to my-proxy-endpoint.customerdomain.com/207.148.248.143:443
```

File: aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/ProgressListenerHttpInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (gpanshu) [inline aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/ProgressListenerHttpInterceptor.kt:2]: 2023

**Review by gpanshu — APPROVED** [review]
nit pick on the license header but otherwise LGTM

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthService.kt [hunk 3]
```diff
@@ -30,8 +30,8 @@ interface AWSCognitoAuthService {
 
                 CognitoIdentityProviderClient {
                     this.region = it.region
-                    this.endpointResolver = it.endpoint?.let { endpoint ->
-                        AWSEndpointResolver(Endpoint(endpoint))
+                    this.endpointProvider = it.endpoint?.let { endpoint ->
+                        AWSEndpointProvider(Endpoint(endpoint))
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthService.kt:34]:
```suggestion
                    this.endpointProvider = it.endpoint?.let { endpoint ->
                        EndpointProvider { Endpoint(endpoint) }
                    }
```

We can get rid of the below `AWSEndpointProvider` altogether.

File: aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/TransferTable.kt [hunk 4]
```diff
@@ -213,6 +216,9 @@ internal class TransferTable {
             if (TABLE_VERSION_8 in (oldVersion + 1)..newVersion) {
                 addVersion8Columns(database)
             }
+            if (TABLE_VERSION_9 in (oldVersion + 1)..newVersion) {
```
Comment (tylerjroach) [inline aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/TransferTable.kt:219]: Can we get a test on this migration? If not in code, at least confirmation v8 db with existing entries was observed to migrate successfully and set default values of existing rows to 0

File: maplibre-adapter/src/main/java/com/amplifyframework/geo/maplibre/http/AWSRequestSignerInterceptor.kt [hunk 5]
```diff
@@ -106,8 +108,8 @@ internal class AWSRequestSignerInterceptor(
         }
 
         val httpUrl = Url(
```
Comment (tylerjroach) [inline maplibre-adapter/src/main/java/com/amplifyframework/geo/maplibre/http/AWSRequestSignerInterceptor.kt:110]: Haven't looked at Kotlin sdk but what is the reason that the port is needed for both the scheme and the port val below.

**Review by tylerjroach — APPROVED** [review]
I approve of the PR but it looks like this will have to be held off until Kotlin SDK team resolves the Polly issue Erica discovered and we can update to that version.

File: aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/TransferTable.kt (hunk 4 above)
Comment (sdhuka) [reply aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/TransferTable.kt:219]: image

**Review by div5yesh — DISMISSED** [review]

File: maplibre-adapter/src/main/java/com/amplifyframework/geo/maplibre/http/AWSRequestSignerInterceptor.kt (hunk 5 above)
Comment (sdhuka) [reply maplibre-adapter/src/main/java/com/amplifyframework/geo/maplibre/http/AWSRequestSignerInterceptor.kt:110]: It port value is not set then it defaults to schema.port
```
public data class Url(
    public val scheme: Scheme,
    public val host: Host,
    public val port: Int = scheme.defaultPort,
    public val path: String = "",
    public val parameters: QueryParameters = QueryParameters.Empty,
    public val fragment: String? = null,
    public val userInfo: UserInfo? = null,
    public val forceQuery: Boolean = false,
    public val encodeParameters: Boolean = true,
)
```

**Review by tylerjroach — APPROVED** [review]

**Review by eeatonaws — COMMENTED** [review]
LGTM, one nit is that the comments for the setUseAcceleratedEndpoint method all say they're for acceleration mode for DownloadFileOptions, which doesn't always match the options class (e.g., StorageUploadFileOptions).

**Review by eeatonaws — APPROVED** [review]

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**